src/common/libpmi/simple_server.c

```diff
@@ -241,8 +243,26 @@ int pmi_simple_server_request (struct pmi_simple_server *pmi,
     }
     /* abort */
     else if (keyval_parse_isword (buf, "cmd", "abort") == 0) {
-        /* FIXME - terminate program */
-        snprintf (resp, sizeof (resp), "\n");
+        unsigned int code;
+        char *msg = "aborted";
+        char error_msg[SIMPLE_KVS_VAL_MAX];
+
+        /*  mpich circa 2014 sends `exit_code` but not `error_msg`.
+         *  Flux implementation sends both.
+         *  Older mpich and derivatives just exit from the task,
+         *   sending nothing.
+         */
+        if (keyval_parse_uint (buf, "exit_code", &code) < 0)
+            goto proto;
+        if (keyval_parse_string (buf,
+                                 "error_msg",
+                                 error_msg,
+                                 sizeof (error_msg)) == 0)
+            msg = error_msg;
+        if (pmi->ops.abort)
+            pmi->ops.abort (pmi->arg, cli, code, msg);
+
+        /*  Abort call above should kill program, o/w continue as before */
```

Comment:
I'm not sure I follow. Shouldn't the *server* be allowed to continue if it needs to do so?

If the server process exits, then the mpich task will get EOF on the PMI fd. Can we simulate that by closing the server's end of the FD if `pmi->ops.abort` returns?

Reply:
> I'm not sure I follow. Shouldn't the server be allowed to continue if it needs to do so?

Yes you're right, sorry, the server should be able to continue.

> If the server process exits, then the mpich task will get EOF on the PMI fd. Can we simulate that by closing the server's end of the FD if pmi->ops.abort returns?

I don't think we want to simulate that as it would let the internal `PMI_Abort()` in mpich return into the post-assertion zombie apocolypse.

I think the bottom line is if the process manager receives an abort request, it's the PM's responsibility to terminate the task.  The task isn't always going to do it on its own.  Therefore, the abort callback should be implemented to avoid a potential hang on abort.

Since the shell does provide the abort callback, I guess this is not an issue, so let's leave the server side code as is.